chat-plugins/tickets.js

```diff
@@ -0,0 +1,487 @@
+			let ticket = tickets[user.userid];
+			if ((ticket && ticket.open) || checkIp(user.latestIp)) {
+				return this.sendReply(`|html|You already have an open ticket, please wait for global staff to respond.${(ticket.userid === user.userid && !user.inRooms.has('help-' + user.userid)) ? ` <button class="button" name="send" value="/join help-${user.userid}">Join ticket room</button>` : ``}`);
+			}
+			let banMsg = checkTicketBanned(user);
+			if (banMsg) return this.errorReply(banMsg);
+			if (Monitor.countTickets(user.latestIp)) return user.popup(`Due to high load, you are limited to creating ${Punishments.sharedIps.has(user.latestIp) ? `50` : `5`} tickets every hour.`);
+			if (!['PM Harrasment', 'Battle Harrasment', 'Chatroom Harrasment', 'Inappropriate Content', 'Inappropriate Nickname', 'Inappropriate Pokemon Nicknames', 'Timerstalling', 'Global Staff Complaint', 'Appeal', 'IP-Appeal', 'Lost Password', 'Room Owner Complaint', 'Other'].includes(target)) return this.parse('/ticket');
+			let upper = false;
+			if (['Lost Password', 'Room Owner Complaint', 'Global Staff Complaint'].includes(target)) upper = true;
+			ticket = {
+				creator: user.name,
+				userid: user.userid,
+				open: true,
+				type: target,
+				created: Date.now(),
+				claimed: null,
+				escalated: upper,
+				ip: user.latestIp,
+			};
+			let helpRoom = Rooms('help-' + user.userid);
+			if (!helpRoom) {
+				helpRoom = Rooms.createChatRoom('help-' + user.userid, '[H] ' + user.userid, {
+					isPersonal: true,
+					isHelp: 'open',
+					isPrivate: 'hidden',
+					modjoin: (upper ? '&' : '%'),
+					auth: {[user.userid]: '+'},
+					introMessage: `<h2 style="margin-top:0">Help Ticket - ${user.name}</h2><p><b>Issue</b>: ${ticket.type}<br />${upper ? `An Upper` : `A Global`} Staff member will be with you shortly.</p>`,
+					staffMessage: `<p><button class="button" name="send" value="/ticket close ${user.userid}">Close Ticket</button> <button class="button" name="send" value="/ticket escalate ${user.userid}">Escalate</button> ${upper ? `` : `<button class="button" name="send" value="/ticket escalate ${user.userid}, upperstaff">Escalate to Upper Staff</button>`}</p><p>To ban this user from creating tickets for two days, please use <code>/ticket ban ${user.userid}</code></p>`,
+				});
+				helpRoom.game = new HelpTicket(helpRoom, ticket);
+			} else {
+				helpRoom.isHelp = 'open';
+				if (helpRoom.expireTimer) clearTimeout(helpRoom.expireTimer);
+				if (upper && helpRoom.modjoin === '%') {
+					// Kick drivers and moderators out
+					helpRoom.modjoin = '&';
+					for (let u in helpRoom.users) {
+						let targetUser = helpRoom.users[u];
+						if (targetUser.isStaff && ['%', '@'].includes(targetUser.group)) targetUser.leaveRoom(helpRoom);
+					}
+				} else if (!upper && helpRoom.modjoin === '&') {
+					helpRoom.modjoin = '%';
+					// TODO clear chat log to remove upper staff only information?
```

Comment:
This would be something log-viewer side. I suppose it could be resolved with a policy of taking sensitive issues to PMs.
Or possibly a different room name (though that would be messy).

Reply:
Yeah, taking sensitive issues to PMs is probably the better idea.

Reply:
How does the log viewer work for permission checking on the Upper Staff? Is there some room type or something that uses? Could this somehow replicate that maybe?

Reply:
Ill add a bold msg to the staff intro if its upper staff only to prevent mistakes.